File: packages/renderer-vue/playground/MathNode.ts
import { defineNode, NodeInterface } from "@baklavajs/core";
import { NumberInterface, SelectInterface } from "../src";
import { setType } from "@baklavajs/interface-types";
import { numberType } from "./interfaceTypes";

export default defineNode({
    type: "MathNode",
    title: "Math",
    inputs: {
        operation: () => new SelectInterface("Operation", "Add", ["Add", "Subtract", "Divide"]).setPort(false),
        number1: () => new NumberInterface("Number", 1).use(setType, numberType),
        number2: () => new NumberInterface("Number", 10).use(setType, numberType),
    },
    outputs: {
        result: () => new NodeInterface<number>("Result", 0).use(setType, numberType),
    },
    calculate({ number1, number2, operation }) {
        let result: number;
        if (operation === "Add") {
            result = number1 + number2;
        }
        else if (operation === "Subtract")
        {
            result = number1 - number2;
        }
        else if (operation == "Divide")
        {
            result = number1 / number2;
        }
        else
        {
            throw new Error(`Unknown operation: ${operation}`);
        }
        return { result };
    },
});

File: BaklavaDependencyEngine/BaklavaDependencyEngine.Core/ExampleNodes/MathNode.cs
using BaklavaDependencyEngine.Core.Models;

namespace BaklavaDependencyEngine.Core.ExampleNodes
{
    public class MathNode : Node
    {
        public enum Operation
        {
            Add,
            Subtract,
            Multiply,
            Divide
        }

        public Operation MathOperation { get; set; } = Operation.Add;

        public MathNode() : base("MathNode", "Math")
        {
        }

        protected override void InitializeInterfaces()
        {
            AddInput("number1", new NodeInterface("Number 1", 0.0));
            AddInput("number2", new NodeInterface("Number 2", 0.0));
            AddOutput("result", new NodeInterface("Result", 0.0, false));
        }

        public override async Task<Dictionary<string, object>> Calculate(
            Dictionary<string, object> inputs,
            CalculationContext context)
        {
            var num1 = Convert.ToDouble(inputs["number1"]);
            var num2 = Convert.ToDouble(inputs["number2"]);
            double result;

            switch (MathOperation)
            {
                case Operation.Add:
                    result = num1 + num2;
                    break;
                case Operation.Subtract:
                    result = num1 - num2;
                    break;
                case Operation.Multiply:
                    result = num1 * num2;
                    break;
                case Operation.Divide:
                    if (Math.Abs(num2) < 0.0001)
                        throw new DivideByZeroException();
                    result = num1 / num2;
                    break;
                default:
                    throw new InvalidOperationException($"Unknown operation: {MathOperation}");
            }

            return await Task.FromResult(new Dictionary<string, object>
            {
                ["result"] = result
            });
        }
    }
}